wp-includes/class-wp-https-ui.php

```diff
@@ -0,0 +1,336 @@
+	 * @return bool Whether the option is true or false.
+	 */
+	public function upgrade_https_sanitize_callback( $raw_value ) {
+		unset( $raw_value );
+		return isset( $_POST[ self::UPGRADE_HTTPS_OPTION ] ); // WPCS: CSRF OK.
+	}
+
+	/**
+	 * Adds a settings field to the 'General Settings' page.
+	 *
+	 * Only add this if the site can support HTTPS,
+	 * but the home and siteurl option values are not HTTPS (WordPress Address and Site Address).
+	 * This UI would not apply if those URLs are already HTTPS.
+	 */
+	public function add_settings_field() {
+		if ( ! $this->wp_https_detection->is_currently_https() ) {
+			add_settings_field(
+				self::HTTPS_SETTING_ID,
+				__( 'HTTPS', 'pwa' ),
+				array( $this, 'render_https_settings' ),
+				self::OPTION_GROUP
+			);
+		}
+	}
+
+	/**
+	 * Renders the HTTPS settings in /wp-admin on the General Settings page.
+	 */
+	public function render_https_settings() {
+		$upgrade_https_value = (bool) get_option( self::UPGRADE_HTTPS_OPTION );
+		$https_more_details  = sprintf(
+			'<a href="%s">%s</a>',
+			__( 'https://make.wordpress.org/support/user-manual/web-publishing/https-for-wordpress/', 'pwa' ),
+			__( 'More details', 'pwa' )
+		);
+
+		?>
+		<p class="description">
+			<?php
+			echo wp_kses_post( sprintf(
+				/* translators: %s: a link for more details */
+				__( 'HTTPS is essential to securing your WordPress site, we strongly suggest upgrading to it. %s', 'pwa' ),
+				$https_more_details
+			) );
+			?>
+		</p>
+		<p>
+			<label><input name="<?php echo esc_attr( self::UPGRADE_HTTPS_OPTION ); ?>" type="checkbox" <?php checked( $upgrade_https_value ); ?> value="<?php echo esc_attr( self::OPTION_CHECKED_VALUE ); ?>"><?php esc_html_e( 'Upgrade to secure connection', 'pwa' ); ?></label>
+		</p>
+		<script>
+			( function ( $ ) {
+				// Move this UI under the Site Address (URL) <tr> on the General Settings page.
+				$( 'input[name=<?php echo esc_attr( self::UPGRADE_HTTPS_OPTION ); ?>]' ).parents( 'tr' ).insertAfter( $( 'label[for=home]' ).parents( 'tr') );
+			} )( jQuery );
+		</script>
+		<?php
+
+		$all_insecure_urls = get_option( WP_HTTPS_Detection::INSECURE_CONTENT_OPTION_NAME );
+		$total_urls_count  = is_array( $all_insecure_urls ) && ! empty( $all_insecure_urls ) ? count( $all_insecure_urls ) : 0;
+
+		// Exit if there is no insecure URL to display.
+		if ( ! $total_urls_count ) {
+			return;
+		}
+
+		$insecure_content_id = 'insecure-content';
+		$show_more_button_id = 'view-urls';
+		$insecure_urls_class = 'insecure-urls';
+		$description         = sprintf(
+			/* translators: %s is a link for more details */
+			__( 'We found content on your site that wasn&#39;t loading correctly over HTTPS. While we will try to fix these links automatically, you might check to be sure your pages work as expected. %s', 'pwa' ),
+			sprintf(
+				'<a href="%s">%s</a>',
+				__( 'https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content/How_to_fix_website_with_mixed_content', 'pwa' ),
```

Comment:
This should have `esc_url()` wrapping.

Reply:
Sure, https://github.com/xwp/pwa-wp/pull/36/commits/2a87852716f186ea3e83d182cd3856837fb2c035 adds `esc_url()`.